sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1902,218 @@ def transform(self, X):
         """
         return _transform_selected(X, self._transform,
                                    self.categorical_features, copy=True)
+
+
+class CountFeaturizer(BaseEstimator, TransformerMixin):
+    """Adds a feature representing each feature value's count in training
+
+    Specifically, for each data point X_i in the dataset X, it will add in
+    a new set of columns count_X_i to the end of X_i where count_X_i is the
+    number of occurences of X_i in the dataset X given the equality indicator
+    'inclusion'.
+
+    If a y argument is given during the fit step, then the
+    count in the transform step will be conditional on the y.
+    The number of columns added will be the number of different values y can
+    take on.
+
+    This preprocessing step is useful when the number of occurences
+    of a particular piece of data is helpful in computing the prediction.
+
+    Parameters
+    ----------
+    inclusion : 'all', list, or numpy.ndarray
+        The inclusion criteria for counting
+
+        - 'all' (default) : Every feature given is counted
+        - collection of indices : Only the given collection of features is
+        counted
+
+    Attributes
+    ----------
+    count_cache_ : defaultdict(int)
+        The counts of each example learned during 'fit'
+
+    y_set_ : list of (index, y) tuples
+        An enumerated set of all unique values y can have
+
+    col_num_X_ : int
+        The number of columns of 'X' learned during 'fit'
```

Comment:
Essentially it's the number of columns of `X` that is put into the fit() function  
Then when transform is called on another `X2`, we verify that `X2` has the same number of columns as `X`, if the number of columns are different, an error is thrown